lib/auth/webauthn/login_test.go

```diff
@@ -621,6 +654,162 @@ func TestCredentialRPID(t *testing.T) {
 	})
 }
 
+func TestLogin_scopeAndReuse(t *testing.T) {
+	// webUser gets a newly registered device and a webID.
+	const webUser = "llama"
+	webIdentity := newFakeIdentity(webUser)
+	webConfig := &types.Webauthn{RPID: "example.com"}
+
+	const webOrigin = "https://example.com"
+	ctx := context.Background()
+
+	// Register a Webauthn device.
+	// Last registration step creates the user webID and adds the new device to
+	// identity.
+	webKey, err := mocku2f.Create()
+	require.NoError(t, err)
+	webKey.PreferRPID = true // Webauthn-registered device
+	webRegistration := &wanlib.RegistrationFlow{
+		Webauthn: webConfig,
+		Identity: webIdentity,
+	}
+	cc, err := webRegistration.Begin(ctx, webUser, false /* passwordless */)
+	require.NoError(t, err)
+	ccr, err := webKey.SignCredentialCreation(webOrigin, cc)
+	require.NoError(t, err)
+	device, err := webRegistration.Finish(ctx, wanlib.RegisterResponse{
+		User:             webUser,
+		DeviceName:       "webauthn1",
+		CreationResponse: ccr,
+	})
+	require.NoError(t, err)
+
+	tests := []struct {
+		name          string
+		challengeExt  mfav1.ChallengeExtensions
+		wantBeginErr  error
+		requiredExt   mfav1.ChallengeExtensions
+		wantFinishErr error
+	}{
+		{
+			name: "NOK scope not satisfied",
+			challengeExt: mfav1.ChallengeExtensions{
+				Scope: mfav1.ChallengeScope_CHALLENGE_SCOPE_ADMIN_ACTION,
+			},
+			requiredExt: mfav1.ChallengeExtensions{
+				Scope: mfav1.ChallengeScope_CHALLENGE_SCOPE_LOGIN,
+			},
+			wantFinishErr: trace.AccessDenied("required scope %q is not satisfied by the given webauthn session with scope %q", mfav1.ChallengeScope_CHALLENGE_SCOPE_LOGIN, mfav1.ChallengeScope_CHALLENGE_SCOPE_ADMIN_ACTION),
```

Comment:
I suggest asserting only part of the message, with ErrorContains, so we are not so sensitive to text changes.

Maybe do `assertFinishErr func(t *testing.T, err error)` and do both assertions in a custom func (`assert.ErrorAs(t, err, &trace.AccessDeniedError{})` plus `assert.ErrorContains(t, err, msg)`).

Same for similar blocks.